src/com/amazon/ion/impl/IonReaderContinuableTopLevelBinary.java

```diff
@@ -70,13 +106,52 @@ public boolean hasNext() {
         throw new UnsupportedOperationException("Not implemented");
     }
 
+    /**
+     * Advances to the next value and attempts to fill it.
+     * @return true if not enough data was available in the stream; otherwise, false.
+     */
+    private boolean nextAndFill() {
+        while (true) {
+            if (!isFillingValue) {
+                if (nextValue() == IonCursor.Event.NEEDS_DATA) {
+                    type = null;
+                    return true;
+                }
+            }
+            isFillingValue = true;
+            event = fillValue();
+            if (event == IonCursor.Event.NEEDS_DATA) {
+                type = null;
+                return true;
+            } else if (event == IonCursor.Event.NEEDS_INSTRUCTION) {
+                // The value was skipped for being too large. Get the next one.
+                isFillingValue = false;
+                continue;
+            }
+            break;
+        }
+        isFillingValue = false;
+        return false;
+    }
+
     @Override
     public IonType next() {
-        IonCursor.Event event = super.nextValue();
-        if (event == IonCursor.Event.NEEDS_DATA) {
-            endStream();
-            type = null;
-            return null;
+        if (!isSlowMode || isNonContinuable || parent != null) {
+            IonCursor.Event event = super.nextValue();
+            if (event == IonCursor.Event.NEEDS_DATA) {
+                if (isNonContinuable) {
+                    endStream();
+                }
+                type = null;
+                return null;
+            } else if (event == IonCursor.Event.NEEDS_INSTRUCTION) {
+                // The value was skipped for being too large.
+                isFillingValue = false;
```

Comment:
I'm having trouble seeing the expected interaction between the caller and the reader state here, and I think the disconnect is between their view of what _has_ been skipped vs this code comment.

If I read this right, the user gets the Type of the value that _will_ be skipped (from their POV), let's say it's a _really_ `BigInteger`. Do they get back `IonInt` from this call (and then what do they do?!?) or a `null`? or I'm just way off here.

Reply:
When a value is skipped due to being too large, nothing about that value is surfaced to the user (but note that they will receive a callback if they provided a callback method in their buffer configuration). That's why this is in a loop; if the value was too large, we continue the loop until we find a value that fits. When such a value is found, that value's type is returned.